src/include/platform/NetworkCommissioning.h

```diff
@@ -0,0 +1,189 @@
+namespace DeviceLayer {
+
+class DLL_EXPORT WiFiNetworkCommissioningDelegate
+{
+public:
+    struct WiFiNetworkInfo
+    {
+        uint8_t mSSID[DeviceLayer::Internal::kMaxWiFiSSIDLength + 1];
+        uint8_t mSSIDLen;
+        uint8_t mCredentials[DeviceLayer::Internal::kMaxWiFiKeyLength];
+        uint8_t mCredentialsLen;
+    };
+
+    class Callback
+    {
+    public:
+        virtual void OnConnectWiFiResult(app::Clusters::NetworkCommissioning::NetworkCommissioningStatus commissioningError,
+                                         CharSpan debugText, int32_t wifiAssociationStatus) = 0;
+
+        /**
+         * Indicates the scan is finished, and accepts a list of networks discovered.
+         * OnScanFinished() must be called in a thread-safe manner with CHIP stack. (e.g. using ScheduleWork or ScheduleLambda)
+         */
+        virtual void
+        OnScanFinished(CHIP_ERROR err, CharSpan debugText,
+                       const Span<app::Clusters::NetworkCommissioning::Structs::WiFiInterfaceScanResult::Type> & network) = 0;
+
+        virtual ~Callback() = default;
+    };
+
+    /**
+     * @brief
+     *   Called to initialtiate a WiFi network scan, exactly one of callback->OnError or callback->OnScanFinished must be called
+     * within timeout.
+     *   The ScanWiFiNetworks has no return values, platforms can call callback->OnError() inside ScanWiFiNetworks.
+     *
+     * @param timeout The timeout of the scan, the scan MUST be finished or terminated within the given timeout.
+     * @param ssid    The interested SSID, the scanning MAY be restricted to to the given SSID.
+     */
+    virtual void ConnectWiFiNetwork(ByteSpan ssid, ByteSpan credentials, Callback * callback) = 0;
+
+    /**
+     * @brief
+     *   Called to initialtiate a WiFi network scan, exactly one of callback->OnError or callback->OnScanFinished must be called
+     * within timeout.
+     *   The ScanWiFiNetworks has no return values, platforms can call callback->OnError() inside ScanWiFiNetworks.
+     *
+     * @param timeout The timeout of the scan, the scan MUST be finished or terminated within the given timeout.
+     * @param ssid    The interested SSID, the scanning MAY be restricted to to the given SSID.
+     */
+    virtual void ScanWiFiNetworks(ByteSpan ssid, Callback * callback) = 0;
+
+    /**
+     * @brief
+     *   Called to retrieve a list of WiFi networks configured. The implementation should fill no more than maxNetworks items in
+     * the networks.
+     *
+     * @param networks The buffer for filling WiFi network credentials.
+     * @param maxNetworks The size of the WiFi network info buffer.
+     * @return The actual number of networks retrieved from platform driver.
+     */
+    virtual size_t GetNetworks(WiFiNetworkInfo * networks, size_t maxNetworks) = 0;
+
+    /**
+     * @brief
+     *   Called to persist the networks.
+     */
+    virtual CHIP_ERROR StoreNetworks(Span<WiFiNetworkInfo> networks) = 0;
```

Comment:
This should be called `CommitNetworks()` and take no arguments. The underlying impl should have kept track of the actual data.

Overall, this interface must delegate the actual storage of all network config to the implementation, rather than expect the caller to maintain all memory/references to the network configs in cluster representation.